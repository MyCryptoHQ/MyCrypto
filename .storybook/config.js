import React from 'react';
import { addDecorator } from '@storybook/react';
import { withA11y } from '@storybook/addon-a11y';
import { withThemesProvider } from 'storybook-addon-styled-component-theme';
import styled from 'styled-components';
import { light, dark } from '@mycrypto/ui';

import { GAU_COLORS } from 'v2/theme';

import 'sass/styles'

const LIGHT_THEME = {
  name: 'Light',
  ...light,
  GAU: {
    COLORS: GAU_COLORS
  }
};

const DARK_THEME = {
  name: 'Dark',
  ...dark,
  GAU: {
    COLORS: GAU_COLORS
  }
};

addDecorator(withThemesProvider([LIGHT_THEME, DARK_THEME]));
addDecorator(withA11y);
