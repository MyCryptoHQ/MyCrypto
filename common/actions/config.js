// @flow

/*** Change Language ***/
export type ChangeLanguageAction = {
  type: 'CONFIG_LANGUAGE_CHANGE',
  value: string
};

export function changeLanguage(sign: string): ChangeLanguageAction {
  return {
    type: 'CONFIG_LANGUAGE_CHANGE',
    value: sign
  };
}

/*** Change Node ***/
export type ChangeNodeAction = {
  type: 'CONFIG_NODE_CHANGE',
  // FIXME $keyof?
  value: string
};

export function changeNode(value: string): ChangeNodeAction {
  return {
    type: 'CONFIG_NODE_CHANGE',
    value
  };
}

/*** Change gas price ***/
export type ChangeGasPriceAction = {
  type: 'CONFIG_GAS_PRICE',
  value: number
};

export function changeGasPrice(value: number): ChangeGasPriceAction {
  return {
    type: 'CONFIG_GAS_PRICE',
    value
  };
}

/*** Union Type ***/
export type ConfigAction =
  | ChangeNodeAction
  | ChangeLanguageAction
  | ChangeGasPriceAction;
