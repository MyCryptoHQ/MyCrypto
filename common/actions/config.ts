/*** Change Language ***/
export interface ChangeLanguageAction {
  type: 'CONFIG_LANGUAGE_CHANGE';
  value: string;
}

export function changeLanguage(sign: string): ChangeLanguageAction {
  return {
    type: 'CONFIG_LANGUAGE_CHANGE',
    value: sign
  };
}

/*** Change Node ***/
export interface ChangeNodeAction {
  type: 'CONFIG_NODE_CHANGE';
  // FIXME $keyof?
  value: string;
}

export function changeNode(value: string): ChangeNodeAction {
  return {
    type: 'CONFIG_NODE_CHANGE',
    value
  };
}

/*** Change gas price ***/
export interface ChangeGasPriceAction {
  type: 'CONFIG_GAS_PRICE';
  value: number;
}

export function changeGasPrice(value: number): ChangeGasPriceAction {
  return {
    type: 'CONFIG_GAS_PRICE',
    value
  };
}

/*** Union Type ***/
export type ConfigAction =
  | ChangeNodeAction
  | ChangeLanguageAction
  | ChangeGasPriceAction;
