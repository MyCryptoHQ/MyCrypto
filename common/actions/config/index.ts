export * from './actionCreators';
export * from './actionTypes';
