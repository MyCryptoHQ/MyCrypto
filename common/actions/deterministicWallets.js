// @flow
import type Big from 'bignumber.js';

export type TokenValues = { [string]: ?Big };

export type DeterministicWalletData = {
  index: number,
  address: string,
  value?: Big,
  tokenValues: TokenValues
};

/*** Get determinstic wallets ***/
export type GetDeterministicWalletsAction = {
  type: 'DW_GET_WALLETS',
  payload: {
    seed: ?string,
    dPath: string,
    publicKey: ?string,
    chainCode: ?string,
    limit: number,
    offset: number
  }
};

export type GetDeterministicWalletsArgs = {
  seed: ?string,
  dPath: string,
  publicKey: ?string,
  chainCode: ?string,
  limit?: number,
  offset?: number
};

export function getDeterministicWallets(
  args: GetDeterministicWalletsArgs
): GetDeterministicWalletsAction {
  const { seed, dPath, publicKey, chainCode, limit, offset } = args;
  return {
    type: 'DW_GET_WALLETS',
    payload: {
      seed,
      dPath,
      publicKey,
      chainCode,
      limit: limit || 5,
      offset: offset || 0
    }
  };
}

/*** Set deterministic wallets ***/
export type SetDeterministicWalletsAction = {
  type: 'DW_SET_WALLETS',
  payload: DeterministicWalletData[]
};

export function setDeterministicWallets(
  wallets: DeterministicWalletData[]
): SetDeterministicWalletsAction {
  return {
    type: 'DW_SET_WALLETS',
    payload: wallets
  };
}

/*** Set desired token ***/
export type SetDesiredTokenAction = {
  type: 'DW_SET_DESIRED_TOKEN',
  payload: ?string
};

export function setDesiredToken(token: ?string): SetDesiredTokenAction {
  return {
    type: 'DW_SET_DESIRED_TOKEN',
    payload: token
  };
}

/*** Set wallet values ***/
export type UpdateDeterministicWalletArgs = {
  address: string,
  value: ?Big,
  tokenValues: ?TokenValues
};

export type UpdateDeterministicWalletAction = {
  type: 'DW_UPDATE_WALLET',
  payload: UpdateDeterministicWalletArgs
};

export function updateDeterministicWallet(
  args: UpdateDeterministicWalletArgs
): UpdateDeterministicWalletAction {
  return {
    type: 'DW_UPDATE_WALLET',
    payload: args
  };
}

/*** Union Type ***/
export type DeterministicWalletAction =
  | GetDeterministicWalletsAction
  | UpdateDeterministicWalletAction
  | SetDesiredTokenAction;
