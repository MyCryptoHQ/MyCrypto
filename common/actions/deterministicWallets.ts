import { BigNumber } from 'bignumber.js';

export interface TokenValues {
  [key: string]: BigNumber;
}

export interface DeterministicWalletData {
  index: number;
  address: string;
  value?: BigNumber;
  tokenValues: TokenValues;
}

/*** Get determinstic wallets ***/
export interface GetDeterministicWalletsAction {
  type: 'DW_GET_WALLETS';
  payload: {
    seed?: string;
    dPath: string;
    publicKey?: string;
    chainCode?: string;
    limit: number;
    offset: number;
  };
}

export interface GetDeterministicWalletsArgs {
  seed?: string;
  dPath: string;
  publicKey?: string;
  chainCode?: string;
  limit?: number;
  offset?: number;
}

export function getDeterministicWallets(
  args: GetDeterministicWalletsArgs
): GetDeterministicWalletsAction {
  const { seed, dPath, publicKey, chainCode, limit, offset } = args;
  return {
    type: 'DW_GET_WALLETS',
    payload: {
      seed,
      dPath,
      publicKey,
      chainCode,
      limit: limit || 5,
      offset: offset || 0
    }
  };
}

/*** Set deterministic wallets ***/
export interface SetDeterministicWalletsAction {
  type: 'DW_SET_WALLETS';
  payload: DeterministicWalletData[];
}

export function setDeterministicWallets(
  wallets: DeterministicWalletData[]
): SetDeterministicWalletsAction {
  return {
    type: 'DW_SET_WALLETS',
    payload: wallets
  };
}

/*** Set desired token ***/
export interface SetDesiredTokenAction {
  type: 'DW_SET_DESIRED_TOKEN';
  payload: string;
}

export function setDesiredToken(token: string): SetDesiredTokenAction {
  return {
    type: 'DW_SET_DESIRED_TOKEN',
    payload: token
  };
}

/*** Set wallet values ***/
export interface UpdateDeterministicWalletArgs {
  address: string;
  value?: BigNumber;
  tokenValues?: TokenValues;
  index?: any;
}

export interface UpdateDeterministicWalletAction {
  type: 'DW_UPDATE_WALLET';
  payload: UpdateDeterministicWalletArgs;
}

export function updateDeterministicWallet(
  args: UpdateDeterministicWalletArgs
): UpdateDeterministicWalletAction {
  return {
    type: 'DW_UPDATE_WALLET',
    payload: args
  };
}

/*** Union Type ***/
export type DeterministicWalletAction =
  | GetDeterministicWalletsAction
  | UpdateDeterministicWalletAction
  | SetDesiredTokenAction
  | SetDeterministicWalletsAction;
