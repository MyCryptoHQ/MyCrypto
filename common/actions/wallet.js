// @flow
import type { IWallet } from 'libs/wallet/IWallet';
import Big from 'bignumber.js';
import { Wei } from 'libs/units';

/*** Unlock Private Key ***/
export type PrivateKeyUnlockParams = {
  key: string,
  password: string
};

export type UnlockPrivateKeyAction = {
  type: 'WALLET_UNLOCK_PRIVATE_KEY',
  payload: PrivateKeyUnlockParams
};

export function unlockPrivateKey(
  value: PrivateKeyUnlockParams
): UnlockPrivateKeyAction {
  return {
    type: 'WALLET_UNLOCK_PRIVATE_KEY',
    payload: value
  };
}

/*** Unlock Keystore File ***/
export type KeystoreUnlockParams = {
  file: string,
  password: string
};

export type UnlockKeystoreAction = {
  type: 'WALLET_UNLOCK_KEYSTORE',
  payload: KeystoreUnlockParams
};

export function unlockKeystore(
  value: KeystoreUnlockParams
): UnlockKeystoreAction {
  return {
    type: 'WALLET_UNLOCK_KEYSTORE',
    payload: value
  };
}

/*** Unlock Mnemonic ***/
export type MnemonicUnlockParams = {
  phrase: string,
  pass: string,
  path: string,
  address: string
};

export type UnlockMnemonicAction = {
  type: 'WALLET_UNLOCK_MNEMONIC',
  payload: MnemonicUnlockParams
};

export function unlockMnemonic(
  value: MnemonicUnlockParams
): UnlockMnemonicAction {
  return {
    type: 'WALLET_UNLOCK_MNEMONIC',
    payload: value
  };
}

/*** Set Wallet ***/
export type SetWalletAction = {
  type: 'WALLET_SET',
  payload: IWallet
};

export function setWallet(value: IWallet): SetWalletAction {
  return {
    type: 'WALLET_SET',
    payload: value
  };
}

/*** Set Balance ***/
export type SetBalanceAction = {
  type: 'WALLET_SET_BALANCE',
  payload: Wei
};

export function setBalance(value: Wei): SetBalanceAction {
  return {
    type: 'WALLET_SET_BALANCE',
    payload: value
  };
}

/*** Set Token Balance ***/
export type SetTokenBalancesAction = {
  type: 'WALLET_SET_TOKEN_BALANCES',
  payload: {
    [string]: Big
  }
};

export function setTokenBalances(payload: {
  [string]: Big
}): SetTokenBalancesAction {
  return {
    type: 'WALLET_SET_TOKEN_BALANCES',
    payload
  };
}

/*** Broadcast Tx ***/
export type BroadcastTxRequestedAction = {
  type: 'WALLET_BROADCAST_TX_REQUESTED',
  payload: {
    signedTx: string
  }
};

export function broadcastTx(signedTx: string): BroadcastTxRequestedAction {
  return {
    type: 'WALLET_BROADCAST_TX_REQUESTED',
    payload: {
      signedTx
    }
  };
}

/*** Union Type ***/
export type WalletAction =
  | UnlockPrivateKeyAction
  | SetWalletAction
  | SetBalanceAction
  | SetTokenBalancesAction
  | BroadcastTxRequestedAction;
