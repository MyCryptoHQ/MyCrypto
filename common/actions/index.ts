import { ConfigAction } from './config';
import { CustomTokenAction } from './customTokens';
import { DeterministicWalletAction } from './deterministicWallets';
import { EnsAction } from './ens';
import { NotificationsAction } from './notifications';
import { RatesAction } from './rates';
import { SwapAction } from './swap';
import { TransactionAction } from './transaction';
import { WalletAction } from './wallet';
export type AllActions =
  | ConfigAction
  | CustomTokenAction
  | DeterministicWalletAction
  | EnsAction
  | NotificationsAction
  | RatesAction
  | SwapAction
  | TransactionAction
  | WalletAction;
