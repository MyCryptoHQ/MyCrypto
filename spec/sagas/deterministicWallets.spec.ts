import { configuredStore } from 'store';
import { INode } from 'libs/nodes/INode';
import { cloneableGenerator } from 'redux-saga/utils';
import { all, apply, fork, put, select } from 'redux-saga/effects';
import RpcNode from 'libs/nodes/rpc';
import { getDesiredToken, getWallets } from 'selectors/deterministicWallets';
import { getTokens } from 'selectors/wallet';
import { getNodeLib } from 'selectors/config';
import * as dWalletActions from 'actions/deterministicWallets';
import { Token } from 'config/data';
import {
  getDeterministicWallets,
  updateWalletValues,
  updateWalletTokenValues
} from 'sagas/deterministicWallets';
import { TokenValue, Wei } from 'libs/units';

// init module
configuredStore.getState();

const genWalletData1 = () => ({
  index: 0,
  address: '0x0',
  value: TokenValue('100'),
  tokenValues: {
    OMG: {
      value: TokenValue('100'),
      decimal: 16
    }
  }
});

const genWalletData2 = () => ({
  index: 1,
  address: '0x1',
  value: TokenValue('100'),
  tokenValues: {
    BAT: {
      value: TokenValue('100'),
      decimal: 16
    }
  }
});

const genBalances = () => [Wei('100'), Wei('200')];

describe('getDeterministicWallets*', () => {
  describe('starting from seed', () => {
    const dWallet = {
      seed:
        '1ba4b713b9cf6f91e8e2eea015fc4e107452fa7d8ade32322207967371e5c0fb93289d4dde94ce13625ecc60279d211b6d677c67f54b9e97c7e68afc9ca1b5ea',
      dPath: "m/44'/60'/0'/0"
    };
    const action = dWalletActions.getDeterministicWallets(dWallet);
    const gen = getDeterministicWallets(action);

    it('should match put snapshot', () => {
      expect(gen.next().value).toMatchSnapshot();
    });

    it('should fork updateWalletValues', () => {
      expect(gen.next().value).toEqual(fork(updateWalletValues));
    });

    it('should fork updateWalletTokenValues', () => {
      expect(gen.next().value).toEqual(fork(updateWalletTokenValues));
    });
  });

  describe('starting from publicKey & chainCode', () => {
    const dWallet = {
      dPath: '',
      publicKey: '02fcba7ecf41bc7e1be4ee122d9d22e3333671eb0a3a87b5cdf099d59874e1940f',
      chainCode: '180c998615636cd875aa70c71cfa6b7bf570187a56d8c6d054e60b644d13e9d3',
      limit: 10,
      offset: 0
    };

    const action = dWalletActions.getDeterministicWallets(dWallet);
    const gen = getDeterministicWallets(action);

    it('should match put snapshot', () => {
      expect(gen.next().value).toMatchSnapshot();
    });

    it('should fork updateWalletValues', () => {
      expect(gen.next().value).toEqual(fork(updateWalletValues));
    });

    it('should fork updateWalletTokenValues', () => {
      expect(gen.next().value).toEqual(fork(updateWalletTokenValues));
    });
  });
});

describe('updateWalletValues*', () => {
  const walletData1 = genWalletData1();
  const walletData2 = genWalletData2();
  const wallets: dWalletActions.DeterministicWalletData[] = [walletData1, walletData2];
  const balances = genBalances();
  const node: INode = new RpcNode('');
  const gen = updateWalletValues();

  it('should select getNodeLib', () => {
    expect(gen.next().value).toEqual(select(getNodeLib));
  });

  it('should select getWallets', () => {
    expect(gen.next(node).value).toEqual(select(getWallets));
  });

  it('should get balance for all wallets', () => {
    expect(gen.next(wallets).value).toEqual(
      all([
        apply(node, node.getBalance, [walletData1.address]),
        apply(node, node.getBalance, [walletData2.address])
      ])
    );
  });

  it('should put updateDeterministicWallet for wallet1', () => {
    expect(gen.next(balances).value).toEqual(
      put(
        dWalletActions.updateDeterministicWallet({
          ...walletData1,
          value: balances[0]
        })
      )
    );
  });

  it('should put updateDeterministicWallet for wallet2', () => {
    expect(gen.next(balances).value).toEqual(
      put(
        dWalletActions.updateDeterministicWallet({
          ...walletData2,
          value: balances[1]
        })
      )
    );
  });

  it('should be done', () => {
    expect(gen.next().done).toEqual(true);
  });
});

describe('updateWalletTokenValues*', () => {
  const walletData1 = genWalletData1();
  const walletData2 = genWalletData2();
  const wallets: dWalletActions.DeterministicWalletData[] = [walletData1, walletData2];
  const node: INode = new RpcNode('');
  const token1: Token = {
    address: '0x2',
    symbol: 'OMG',
    decimal: 16
  };
  const token2: Token = {
    address: '0x3',
    symbol: 'BAT',
    decimal: 16
  };
  const tokens = [token1, token2];
  const tokenBalances = [TokenValue('100'), TokenValue('200')];
  const desiredToken = 'OMG';
  const data = {} as any;
  data.gen = cloneableGenerator(updateWalletTokenValues)();

  it('should select getDesiredToken', () => {
    expect(data.gen.next().value).toEqual(select(getDesiredToken));
  });

  it('should return if desired token is falsey', () => {
    data.clone1 = data.gen.clone();
    data.clone1.next();
    expect(data.clone1.next().done).toEqual(true);
  });

  it('should select getTokens', () => {
    data.clone2 = data.gen.clone();
    expect(data.gen.next(desiredToken).value).toEqual(select(getTokens));
  });

  it('should return if desired token is not amongst tokens', () => {
    data.clone2.next('fakeDesiredToken');
    expect(data.clone2.next(tokens).done).toEqual(true);
  });

  it('should select getNodeLib', () => {
    expect(data.gen.next(tokens).value).toEqual(select(getNodeLib));
  });

  it('should select getWallets', () => {
    expect(data.gen.next(node).value).toEqual(select(getWallets));
  });

  it('should match snapshot of wallet token balances', () => {
    expect(data.gen.next(wallets).value).toMatchSnapshot();
  });

  it('should match snapshot for put wallet1 update', () => {
    expect(data.gen.next(tokenBalances)).toMatchSnapshot();
  });

  it('should match snapshot for put wallet2 update', () => {
    expect(data.gen.next()).toMatchSnapshot();
  });

  it('should be done', () => {
    expect(data.gen.next().done).toEqual(true);
  });
});
