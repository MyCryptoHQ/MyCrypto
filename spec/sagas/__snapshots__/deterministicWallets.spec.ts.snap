// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getDeterministicWallets* starting from publicKey & chainCode should match put snapshot 1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "payload": Array [
        Object {
          "address": "0xa3fA2e024bf9964bb64bDf0afdCa8E3c374a6e41",
          "index": 0,
          "tokenValues": Object {},
        },
        Object {
          "address": "0x9bbC617740413CDCc41fB69901b252e20fd6Ef61",
          "index": 1,
          "tokenValues": Object {},
        },
        Object {
          "address": "0xE5297818aEee60385306b7087B024af20524C0FD",
          "index": 2,
          "tokenValues": Object {},
        },
        Object {
          "address": "0x111789ac11B69fE7EbC307c908Efe7677fd347A2",
          "index": 3,
          "tokenValues": Object {},
        },
        Object {
          "address": "0x315857914bEd907e0Cf33b5883e599dD6ACc45d2",
          "index": 4,
          "tokenValues": Object {},
        },
        Object {
          "address": "0xEA316E5BFd9FDeCD81489929ae56DBE6ffaDD22C",
          "index": 5,
          "tokenValues": Object {},
        },
        Object {
          "address": "0x5De67797cEeCeD707A6868bbE96331ED6f811F06",
          "index": 6,
          "tokenValues": Object {},
        },
        Object {
          "address": "0xd0cb2Fe67f66DF4D6610E99dDE2ad1B0B5bf4054",
          "index": 7,
          "tokenValues": Object {},
        },
        Object {
          "address": "0xC7D3A4101bf3Ac1A5aE898784F80D2F9A1E67B62",
          "index": 8,
          "tokenValues": Object {},
        },
        Object {
          "address": "0x12228Ec03f6aFd6Dad66ceD9C8B9762b3080e319",
          "index": 9,
          "tokenValues": Object {},
        },
      ],
      "type": "DW_SET_WALLETS",
    },
    "channel": null,
  },
}
`;

exports[`getDeterministicWallets* starting from seed should match put snapshot 1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "payload": Array [
        Object {
          "address": "0x2e516E79F439469AA3DD43a93429ee45cBeb77Aa",
          "index": 0,
          "tokenValues": Object {},
        },
        Object {
          "address": "0x4aE493688184D612aF70C0B04aA7f2C8eE03a1dE",
          "index": 1,
          "tokenValues": Object {},
        },
        Object {
          "address": "0xbd208C37C747f784EFfA379aA417d867216Ace19",
          "index": 2,
          "tokenValues": Object {},
        },
        Object {
          "address": "0x85FF91929287F6cdC801280EAf8fa899e22a02F6",
          "index": 3,
          "tokenValues": Object {},
        },
        Object {
          "address": "0xDF5acC09CDf1f966CEc6eFAB3523804E1de65c7f",
          "index": 4,
          "tokenValues": Object {},
        },
      ],
      "type": "DW_SET_WALLETS",
    },
    "channel": null,
  },
}
`;

exports[`updateWalletTokenValues* should match snapshot for put wallet1 update 1`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "PUT": Object {
      "action": Object {
        "payload": Object {
          "address": "0x0",
          "index": 0,
          "tokenValues": Object {
            "OMG": Object {
              "decimal": 16,
              "value": "64",
            },
          },
          "value": "64",
        },
        "type": "DW_UPDATE_WALLET",
      },
      "channel": null,
    },
  },
}
`;

exports[`updateWalletTokenValues* should match snapshot for put wallet2 update 1`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "PUT": Object {
      "action": Object {
        "payload": Object {
          "address": "0x1",
          "index": 1,
          "tokenValues": Object {
            "BAT": Object {
              "decimal": 16,
              "value": "64",
            },
            "OMG": Object {
              "decimal": 16,
              "value": "c8",
            },
          },
          "value": "64",
        },
        "type": "DW_UPDATE_WALLET",
      },
      "channel": null,
    },
  },
}
`;

exports[`updateWalletTokenValues* should match snapshot of wallet token balances 1`] = `
Object {
  "@@redux-saga/IO": true,
  "ALL": Array [
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          "0x0",
          Object {
            "address": "0x2",
            "decimal": 16,
            "symbol": "OMG",
          },
        ],
        "context": RpcNode {
          "client": RPCClient {
            "batch": [Function],
            "call": [Function],
            "createHeaders": [Function],
            "decorateRequest": [Function],
            "endpoint": "",
            "headers": Object {},
          },
          "requests": RPCRequests {},
        },
        "fn": [Function],
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          "0x1",
          Object {
            "address": "0x2",
            "decimal": 16,
            "symbol": "OMG",
          },
        ],
        "context": RpcNode {
          "client": RPCClient {
            "batch": [Function],
            "call": [Function],
            "createHeaders": [Function],
            "decorateRequest": [Function],
            "endpoint": "",
            "headers": Object {},
          },
          "requests": RPCRequests {},
        },
        "fn": [Function],
      },
    },
  ],
}
`;
